recipes-openxt/xenclient-tpm-scripts/xenclient-tpm-scripts/tpm-functions

```diff
@@ -619,213 +621,295 @@ tpm2_write_tboot_policy() {
+# stderr and nothing on stdout.
 tpm_unseal() {
     local root=""
-    local hashalg="sha1"
+    local hashalg="$(tpm_get_hash_algorithm)"
     local pcrs=""
-    local pcr_params=""
     local OPTIND
     local opt
+    # We capture stderr but let stdout untouched using that variable.
+    # The usual sleight of hand goes:
+    #    { stderr="$(cmd args 2>&1 1>&3 3>&-)"; } 3>&1
+    local stderr
 
-    while getopts ":a:r:p:" opt; do
+    while getopts ":a:r:" opt; do
         case "${opt}" in
-            a)
-                hashalg=${OPTARG}
-            ;;
-            r)
-                root=${OPTARG}
-            ;;
-            p)
-                pcrs="${pcrs} ${OPTARG}"
-            ;;
+            a) hashalg="${OPTARG}" ;;
+            r) root="${OPTARG}" ;;
         esac
     done
     shift $((OPTIND-1))
 
-    local tss="${root}/${1}"
+    local tss="${root}/$1"
+    local pcr_opts="$(tpm_pcrs_to_opts $(cat "${tss}.pcrs"))"
 
-    is_tpm_2_0
-    if [ $? -eq 0 ]; then
-        for p in ${pcrs}; do
-            pcr_params="${pcr_params} -r ${p}"
-        done
-
-        pcr_bank_exists "${hashalg}"
-        [ $? -eq 0 ] || return 1
+    if is_tpm_2_0; then
+        if ! pcr_bank_exists "${hashalg}"; then
+            echo "tpm_seal: Requested algorithm \"${hashalg}\" does not "\
+                "exist or is not supported." >&2
+            return 1
+        fi
 
-        case $hashalg in
+        case "${hashalg}" in
         sha1)
-            tpm2_unsealdata -H ${OXT_HANDLE_SHA1} -n "${tss}.sha1" \
-                            -u "${tss}.pub.sha1" -g ${TPM_ALG_SHA1} \
-                            $( < ${tss}.pcrs ) 2>/dev/null
-            return $?
+            { stderr="$(tpm2_unsealdata -H "${OXT_HANDLE_SHA1}" \
+                -n "${tss}.sha1" -u "${tss}.pub.sha1" \
+                -g "${TPM_ALG_SHA1}" ${pcr_opts} 2>&1 1>&3 3>&-)"; } 3>&1
+            if [ "$?" -ne 0 ]; then
+                echo "tpm_unseal: Failed to unseal data (TPM 2.0) "\
+                    "using sha1 algorithm:" >&2
+                echo "${stderr}" >&2
+                return 1
+            fi
         ;;
         sha256)
-            tpm2_unsealdata -H ${OXT_HANDLE_SHA256} -n "${tss}.sha256" \
-                            -u "${tss}.pub.sha256" -g ${TPM_ALG_SHA256} \
-                            $( < ${tss}.pcrs ) 2>/dev/null
-            return $?
+            { stderr="$(tpm2_unsealdata -H "${OXT_HANDLE_SHA256}" \
+                -n "${tss}.sha256" -u "${tss}.pub.sha256" \
+                -g "${TPM_ALG_SHA256}" ${pcr_opts} 2>&1 1>&3 3>&-)"; } 3>&1
+            if [ "$?" -ne 0 ]; then
+                echo "tpm_unseal: Failed to unseal data (TPM 2.0) "\
+                    "using sha256 algorithm:" >&2
+                echo "${stderr}" >&2
+                return 1
+            fi
         ;;
         *)
+            echo "tpm_unseal: Algorithm \"${hashalg}\" is not supported." >&2
```

Comment:
Since the calls are basically identical, you could refactor to use variables in a single tpm_unseal call.  Something like:
```
case ${hashalg} in
    sha1)
        handle=$OXT_HANDLE_SHA1
        alg=$TPM_ALG_SHA1
        ;;
    sha256)
        handle=${OXT_HANDLE_SHA256}
        alg=${TPM_ALG_SHA256}
        ;;
    *) error "tpm_unseal Alg $hashalg unsupported"
        ;;
esac
{ stderr="$(tpm2_unsealdata -H "${handle}" \
    -n "${tss}.${hashalg}" -u "${tss}.pub.${hashalg}" \
    -g "${alg}" ${pcr_opts} 2>&1 1>&3 3>&-)"; } 3>&1
```

Reply:
... and the same for `tpm_seal` of course.